python/lsst/daf/butler/core/serverModels.py

```diff
@@ -0,0 +1,286 @@
+        return expression
+
+    @classmethod
+    def from_expression(cls, expression: Any) -> ExpressionQueryParameter:
+        """Convert a standard dataset type expression to wire form."""
+        if expression is ...:
+            return cls()
+
+        expressions = iterable(expression)
+        params: Dict[str, List[str]] = {"glob": [], "regex": []}
+        for expression in expressions:
+            if expression is ...:
+                # This matches everything
+                return cls()
+
+            if isinstance(expression, re.Pattern):
+                params["regex"].append(expression.pattern)
+            elif isinstance(expression, str):
+                params["glob"].append(expression)
+            elif hasattr(expression, "name"):
+                params["glob"].append(expression.name)
+            else:
+                raise ValueError(f"Unrecognized type given to expression: {expression!r}")
+
+        # Clean out empty dicts.
+        for k in list(params):
+            if not params[k]:
+                del params[k]
+
+        return cls(**params)
+
+
+class DatasetsQueryParameter(ExpressionQueryParameter):
+    """Represents a specification for a dataset expression query.
+
+    This differs from the standard expression query in that an empty
+    expression will return `None` rather than ``...``.
+    """
+
+    _allow_ellipsis: ClassVar[bool] = False
+
+
+# Shared field definitions
+Where = Field(
+    None,
+    title="String expression similar to a SQL WHERE clause.",
+    example="detector = 5 AND instrument = 'HSC'",
+)
+Collections = Field(
+    None,
+    title="An expression that identifies the collections to search.",
+)
+Datasets = Field(
+    None,
+    title="An expression that identifies dataset types to search (must not match all datasets).",
+)
+OptionalDimensions = Field(
+    None,
+    title="Relevant dimensions to include.",
+    example=["detector", "physical_filter"],
+)
+Dimensions = Field(
+    ...,
+    title="Relevant dimensions to include.",
+    example=["detector", "physical_filter"],
+)
+DataId = Field(
+    None,
+    title="Data ID to constrain the query.",
+)
+FindFirst = Field(
+    False,
+    title="Control whether only first matching dataset ref or type is returned.",
+)
+Components = Field(
+    None,
+    title="Control how expressions apply to components.",
+)
+Bind = Field(
+    None,
+    title="Mapping to use to inject values into the WHERE parameter clause.",
+)
+Check = Field(
+    True,
+    title="Control whether to check the query for consistency.",
+)
+KeywordArgs = Field(
+    None,
+    title="Additional parameters to use when standardizing the supplied data ID.",
+)
+
+
+class QueryBaseModel(BaseModel):
+    """Base model for all query models."""
+
+    @validator("keyword_args", check_fields=False)
+    def check_keyword_args(cls, v, values) -> Optional[SimpleDataId]:  # type: ignore # noqa: N805
```

Comment:
Nit: I tend to name pydantic validators with a leading underscore since they're not intended to be part of the public interface of the class.